File: app/src/main/java/humbleactivity/app/ui/ChainComposerActivity.java
package humbleactivity.app.ui;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import com.jakewharton.rxbinding.view.RxView;

import butterknife.BindView;
import butterknife.ButterKnife;
import humbleactivity.app.HumbleApplication;
import humbleactivity.app.R;
import humbleactivity.app.data.Filter;
import kotlin.Unit;
import kotlin.collections.CollectionsKt;
import rx.subscriptions.CompositeSubscription;

import javax.inject.Inject;

public class ChainComposerActivity extends Activity {
    @Inject ChainComposer presenter;
    private CompositeSubscription subscriptions;

    @BindView(R.id.availables) ListView availablesView;
    ArrayAdapter<String> availablesAdapter;
    @BindView(R.id.chain) ListView chainView;
    ArrayAdapter<String> chainAdapter;
    @BindView(R.id.refresh) View refreshView;
    @BindView(R.id.add_to_chain) View addToChainView;
    @BindView(R.id.remove_from_chain) View removeFromChainView;
    @BindView(R.id.move_down) View moveDownView;
    @BindView(R.id.move_up) View moveUpView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        HumbleApplication.get(this).component().inject(this);

        setContentView(R.layout.activity_chain_composer);
        ButterKnife.bind(this);

        subscriptions = new CompositeSubscription();

        availablesAdapter = new ArrayAdapter<>(this, R.layout.view_filter_item);
        availablesView.setAdapter(availablesAdapter);
        chainAdapter = new ArrayAdapter<>(this, R.layout.view_filter_item);
        chainView.setAdapter(chainAdapter);

        presenter.availables().subscribe(filters -> {
            availablesAdapter.clear();
            availablesAdapter.addAll(CollectionsKt.map(filters, Filter::getName));
        });
        presenter.chain().subscribe(filters -> {
            chainAdapter.clear();
            chainAdapter.addAll(CollectionsKt.map(filters, Filter::getName));
        });
        presenter.chainCursorMove().subscribe(position -> {
            chainView.setItemChecked(position, true);
        });
        subscriptions.add(presenter.loadError().subscribe(message -> Toast.makeText(ChainComposerActivity.this, message, Toast.LENGTH_LONG).show()));

        RxView.clicks(refreshView).map(ignore -> Unit.INSTANCE).subscribe(presenter.onRefresh());
        RxView.clicks(addToChainView).map(ignore -> availablesView.getCheckedItemPosition())
                .filter(p -> p >= 0 && p < availablesView.getCount())
                .subscribe(presenter.onAddToChain());
        RxView.clicks(removeFromChainView).map(ignore -> chainView.getCheckedItemPosition())
                .filter(p -> p >= 0 && p < chainView.getCount())
                .subscribe(presenter.onRemoveFromChain());
        RxView.clicks(moveUpView).map(ignore -> chainView.getCheckedItemPosition())
                .filter(p -> p >= 1 && p < chainView.getCount())
                .subscribe(presenter.onMoveUp());
        RxView.clicks(moveDownView).map(ignore -> chainView.getCheckedItemPosition())
                .filter(p -> p >= 0 && p < chainView.getCount() - 1)
                .subscribe(presenter.onMoveDown());

        presenter.initialize();
    }

    @Override
    protected void onDestroy() {
        presenter.dispose();
        subscriptions.unsubscribe();
        super.onDestroy();
    }
}

File: app/src/main/java/humbleactivity/app/ui/ChainComposerActivity.kt
package humbleactivity.app.ui

import android.app.Activity
import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.Toast
import com.jakewharton.rxbinding2.view.RxView
import humbleactivity.app.HumbleApplication
import humbleactivity.app.R
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_chain_composer.*
import javax.inject.Inject

class ChainComposerActivity : Activity() {
    @Inject
    lateinit var presenter: ChainComposer
    private lateinit var subscriptions: CompositeDisposable

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        HumbleApplication.get(this).component().inject(this)

        setContentView(R.layout.activity_chain_composer)
        val availablesAdapter = ArrayAdapter<String>(this, R.layout.view_filter_item)
        availables.adapter = availablesAdapter
        val chainAdapter = ArrayAdapter<String>(this, R.layout.view_filter_item)
        chain.adapter = chainAdapter

        subscriptions = CompositeDisposable()
        //Output signals
        subscriptions.add(presenter.availables
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { filters ->
                    availablesAdapter.clear()
                    availablesAdapter.addAll(filters.map { it.name })
                })
        subscriptions.add(presenter.chain
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { filters ->
                    chainAdapter.clear()
                    chainAdapter.addAll(filters.map { it.name })
                })
        subscriptions.add(presenter.chainCursor
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { position ->
                    chain.setItemChecked(position, true)
                })
        subscriptions.add(presenter.loadError
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { message ->
                    Toast.makeText(this, message, Toast.LENGTH_LONG).show()
                })
        //Input signals
        subscriptions.add(RxView.clicks(refresh).map { Unit }
                .subscribe(presenter.refresh))
        subscriptions.add(RxView.clicks(add_to_chain).map { availables.checkedItemPosition }
                .filter { p -> p >= 0 && p < availables.count }
                .subscribe(presenter.addToChain))
        subscriptions.add(RxView.clicks(remove_from_chain).map { chain.checkedItemPosition }
                .filter { p -> p >= 0 && p < chain.count }
                .subscribe(presenter.removeFromChain))
        subscriptions.add(RxView.clicks(move_up).map { chain.checkedItemPosition }
                .filter { p -> p >= 1 && p < chain.count }
                .subscribe(presenter.moveUp))
        subscriptions.add(RxView.clicks(move_down).map { chain.checkedItemPosition }
                .filter { p -> p >= 0 && p < chain.count - 1 }
                .subscribe(presenter.moveDown))

        presenter.initialize()
    }

    override fun onDestroy() {
        subscriptions.dispose()
        super.onDestroy()
    }
}
